### PR TITLE
Fix zsvjmp for ppc64le using VRSAVE and SP+12 reserved area

### PR DESCRIPTION
**Pull request by gromero**

### CODE REVIEW

File: zsvjmp-ppc64le.S [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# before returnin and if any function tries to use it ahead it's correctly set
+# as "true".
+
+#define r1  %r1
+#define r3  %r3
+#define r4  %r4
+#define r10 %r10
+#define r11 %r11
+#define r14 %r14
+#define cr0 0
```
Comment (racardoso) [inline zsvjmp-ppc64le.S]: This `#define` is not necessary.

File: zsvjmp-ppc64le.S [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+# Linux and is so set by the kernel to 0xffffffff (-1) since a few libc function
+# use just as boolean. However in this case VRSAVE value is destroy just before
+# calling __sigsetjmp that does not rely on VRSAVE even as boolean. __sigsetjmp
+# just saves the VRSAVE in the context buffer. Once zdojmp triggers the context
+# restoration it restores the VRSAVE incorrect value as well (it still contains
+# the LR LSB value), but once again VRSAVE is fixed by setting it again to -1
+# before returnin and if any function tries to use it ahead it's correctly set
+# as "true".
+
+#define r1  %r1
```
Comment (racardoso) [inline zsvjmp-ppc64le.S]: why not use `%rn` directly on asm code? I understand that this format is more readable than something like `xor 11,11,11` but why create those defines?

File: zsvjmp-ppc64le.S [hunk 3]
```diff
@@ -0,0 +1,61 @@
+
+# Copyright 2018 Gustavo Romero, Rogerio Cardoso, Breno Leitao, IBM Corporation.
+
+	.file "zsvjmp.s"
+	.abiversion 2
+	.section ".text"
+	.align 2
+	.globl zsvjmp_
+	.type zsvjmp_, @function
+
```
Comment (racardoso) [inline zsvjmp-ppc64le.S:10]: This is a really long comment. I suggest to change it to something like:

> Use VRSAVE (unused in Linux) and stack reserved area (SP+12) to save the return address (LR)  which is lost after `bl __sigsetjmp`. External functions should be called using a  `bl/nop` pair, which is turned into a  `plt_call stub` by the linker, ppc64 doesn't  support `b function@plt` anymore. 

And write a comment before line [56](https://github.com/leitao/zsvjmp/pull/2/commits/c4722c4e174890d5fa7e830cd92a2154292510ad#diff-503b9d4c8e36e4be9f5b6278b6a6eb84R56) about why restore VRSAVE:

> VRSAVE is saved on the buffer. Some few libc can use it as a boolean as an optimization so it needs to be restored to -1 (VRSAVE value is always -1).

**Review by racardoso — COMMENTED** [review]
LGTM with those small changes.

File: zsvjmp-ppc64le.S (hunk 1 above)
Comment (racardoso) [reply zsvjmp-ppc64le.S]: `#define crc0 0` is not necessary.

File: zsvjmp-ppc64le.S [hunk 4]
```diff
@@ -8,53 +8,40 @@
...
+	std   %r4,  0(%r3)      # *(buf + 0) = status
+	addi  %r3, %r3, 8       # r3 = buf + 8
+	mr    %r4, %r11         # r4 = 0
 
-	mflr  r11	      # save LR
-	mtvrsave r11          # LR LSB to VRSAVE
-	srdi r11, r11, 32     # LR MSB to SP+12 (reserved)
-	stw  r11, 12(r1)
+	mflr  %r11	        # save LR
+	mtvrsave %r11           # LR LSB to VRSAVE
```
Comment (racardoso) [inline zsvjmp-ppc64le.S]: Trailing white space lines 26, 27.

**Review by racardoso — APPROVED** [review]
LGMT.

**Comment by gromero** [comment]
@racardoso Thanks for reviewing it.

**Comment by gromero** [comment]
@olebole Please, check if that solution fits your purpose. You can download it as a patchset from https://patch-diff.githubusercontent.com/raw/leitao/zsvjmp/pull/2.patch Thank you.

**Comment by gromero** [comment]
This must look like:
```
gromero@gromero16:~/git/zsvjmp$ make clean
rm -f *.o jmptest zzdebug *.a
gromero@gromero16:~/git/zsvjmp$ make 
gcc -g    -c -o zdojmp.o zdojmp.c
gcc -g    -c -o zsvjmp-ppc64le.o zsvjmp-ppc64le.S
ar cr libzsvjmp.a zdojmp.o zsvjmp-ppc64le.o
gromero@gromero16:~/git/zsvjmp$ make test
gfortran -ff2c -g  -c -o jmptest.o jmptest.f
gfortran -o jmptest jmptest.o -L. -lzsvjmp
gcc -g    -c -o zzdebug.o zzdebug.c
gcc -g -g -o zzdebug zzdebug.o -L. -lzsvjmp
./zzdebug
Status = 0, step = 0
Calling zdojmp
Status = 1, step = 1
All OK
./jmptest
 Status =                     0 step =                     0
 Calling zdojmp
 Status =                     1 step =                     1
 All OK
STOP 0
```

**Comment by gromero** [comment]
@leitao @racardoso By reading the glibc's implementation of sigsetjmp/setjmp I could not find any place that is taking care of saving the full stack content. So one could be taken to believe that in some particular situations after the `zdojmp()` invocation this load from stack could load a corrupted value https://github.com/leitao/zsvjmp/pull/2/files/b6e63c6f6accd18eda6f283428f0d7ecb4afa3a0#diff-503b9d4c8e36e4be9f5b6278b6a6eb84R34 However that's not true because it is forbidden to do a `longjmp()` if the function that performed the call to `sigsetjmp/setjmp` already returned (the `main()` function in the  test-case found in `zzdebug.c` called the `segsetjmp` and it cannot return without the program termination, but think of nested calls, like `main()` calling `a()` and `a()` calling `segsetjmp/setjmp`. In that case it would be illegal to call `longjmp()` in `main()` after we returned from `a()`). Thus to sum up, it's not that stack content is save for further use and then restored by `longjmp()`, but that the frame in fact is never destroyed. For instance, [1] clearly says: "Ensure that the function that calls the setjmp function does not return before you call the corresponding longjmp function. Calling the longjmp function after the function calling setjmp returns causes unpredictable program behavior." Also it's in accordance to what I see in glibc's `setjmp` code. Finally I took a look at POSIX specification but it was very unclear about it...

[1] https://www.ibm.com/support/knowledgecenter/en/SSB23S_1.1.0.12/gtpc2/cpp_setjmp.html

**Comment by olebole** [comment]
I can confirm that this PR works nicely within the original software package (IRAF)! So, I would like to include it in my next upload of the package. One question here: I would like to distribute the file under the original [IRAF license](https://github.com/iraf/iraf-v216/blob/9590f45760a4791f3305407fb51c87f1282b32be/local/COPYRIGHTS) (by keeping you as the author ofcourse). Can you confirm that this is OK for you?
A small question aside, for some unspecified future: would the assembler file work for the big endian as well?
Thank you very much for your extensive research and help here! I am very happy to see the package compiling and running under 64-bit PowerPC!

**Comment by gromero** [comment]
@olebole Hi. Thanks for confirming it works for you on LE and for maintaining the code for PPC64. The assembly code works on BE as well. The ABI v1 does not mention the reserved area on ABI's text (contrary to ABI v2) but to the best of my knowledge specific code that deals with CR saving in the stack always performs a `stw` and never a `std` thus reserved area is kept untouched as in LE. VRSAVE is the same. So you just have to change the ABI version to 1 to satisfy the linker. Please take a look at commit https://github.com/leitao/zsvjmp/pull/2/commits/b5d4308181d67c643d3e60c02e60766ef7bc1398 for example. It should goes as:
```
[gromero@gromero22 zsvjmp]$ make clean
rm -f *.o jmptest zzdebug *.a
[gromero@gromero22 zsvjmp]$ make
gcc -g    -c -o zdojmp.o zdojmp.c
gcc -g    -c -o zsvjmp-ppc64.o zsvjmp-ppc64.S
ar cr libzsvjmp.a zdojmp.o zsvjmp-ppc64.o
[gromero@gromero22 zsvjmp]$ make test
gfortran -ff2c -g  -c -o jmptest.o jmptest.f
gfortran -o jmptest jmptest.o -L. -lzsvjmp
gcc -g    -c -o zzdebug.o zzdebug.c
gcc -g -g -o zzdebug zzdebug.o -L. -lzsvjmp
./zzdebug
Status = 0, step = 0
Calling zdojmp
Status = 1, step = 1
All OK
./jmptest
 Status =                     0 step =                     0
 Calling zdojmp
 Status =                     1 step =                     1
 All OK
STOP 0
[gromero@gromero22 zsvjmp]$ uname -a
Linux gromero22 3.10.0-514.el7.ppc64 #1 SMP Wed Oct 19 11:30:41 EDT 2016 ppc64 ppc64 ppc64 GNU/Linux
```
Regarding the distribution question I have to defer it to @leitao who knows better about it.

**Comment by gromero** [comment]
@olebole Please give me a few weeks or so to confirm if it's ok to distribute the file in the current state. Thank you.

**Pull request merged**